src/transport/SecureSessionMgr.cpp

```diff
@@ -111,54 +120,76 @@ CHIP_ERROR SecureSessionMgr::SendMessage(PayloadHeader & payloadHeader, NodeId p
             .SetEncryptionKeyID(state->GetLocalKeyID()) //
             .SetPayloadLength(static_cast<uint16_t>(payloadLength));
         packetHeader.GetFlags().Set(Header::FlagValues::kSecure);
 
         ChipLogProgress(Inet, "Sending msg from %llu to %llu", mLocalNodeId, peerNodeId);
 
-        VerifyOrExit(msgBuf->EnsureReservedSize(headerSize), err = CHIP_ERROR_NO_MEMORY);
+        // Skip encryption process if the packet is resent by CRMP
+        if (!isResend)
+        {
+            VerifyOrExit(msgBuf->EnsureReservedSize(headerSize), err = CHIP_ERROR_NO_MEMORY);
+
+            msgBuf->SetStart(msgBuf->Start() - headerSize);
+            data     = msgBuf->Start();
+            totalLen = msgBuf->TotalLength();
 
-        msgBuf->SetStart(msgBuf->Start() - headerSize);
-        data     = msgBuf->Start();
-        totalLen = msgBuf->TotalLength();
+            err = payloadHeader.Encode(data, totalLen, &actualEncodedHeaderSize);
+            SuccessOrExit(err);
 
-        err = payloadHeader.Encode(data, totalLen, &actualEncodedHeaderSize);
-        SuccessOrExit(err);
+            err = state->GetSecureSession().Encrypt(data, totalLen, data, packetHeader, mac);
+            SuccessOrExit(err);
+            err = mac.Encode(packetHeader, &data[totalLen], kMaxTagLen, &taglen);
+            SuccessOrExit(err);
 
-        err = state->GetSecureSession().Encrypt(data, totalLen, data, packetHeader, mac);
-        SuccessOrExit(err);
+            VerifyOrExit(CanCastTo<uint16_t>(totalLen + taglen), err = CHIP_ERROR_INTERNAL);
+            msgBuf->SetDataLength(static_cast<uint16_t>(totalLen + taglen), nullptr);
 
-        err = mac.Encode(packetHeader, &data[totalLen], kMaxTagLen, &taglen);
-        SuccessOrExit(err);
+            ChipLogDetail(Inet, "Secure transport transmitting msg %u after encryption", msgId);
+        }
 
-        VerifyOrExit(CanCastTo<uint16_t>(totalLen + taglen), err = CHIP_ERROR_INTERNAL);
-        msgBuf->SetDataLength(static_cast<uint16_t>(totalLen + taglen), nullptr);
+        // The Start pointer of msgBuf points to the beginning of the encrypted header, and the DataLength()
+        // contains both the encrypted header and encrypted data. Locally store the start and length and rewind
+        // the start and len back to this state for re-transmission
+        p   = msgBuf->Start();
+        len = msgBuf->DataLength();
 
-        ChipLogDetail(Inet, "Secure transport transmitting msg %u after encryption", state->GetSendMessageIndex());
+        PacketBuffer * buffer = msgBuf.Release_ForNow();
+        err                   = mTransportMgr->SendMessage(packetHeader, state->GetPeerAddress(), buffer);
 
-        err = mTransportMgr->SendMessage(packetHeader, state->GetPeerAddress(), msgBuf.Release_ForNow());
+        // Reset the msgBuf start pointer and data length after sending in case for retransmition
+        buffer->SetStart(p);
+        buffer->SetDataLength(len);
```

Comment:
This isn't great, because conceptually `mTransportMgr->SendMessage(packetHeader, state->GetPeerAddress(), buffer);` calls `Free` on the buffer, so `buffer` might become a dangling pointer.

Now the new caller being added in this PR knows this buffer is not dangling because it happens to have a _different_ handle to it (via the retransmit table), but that's not obvious from inside this function and nothing about this function guarantees that it will always be true.  And I don't think it's true for other callers of this function, for which this code will in fact be doing dangling-pointer dereferences.

Fundamentally, we only have two options for this part:

1. Retain an extra handle across the `mTransportMgr->SendMessage()` call so we can do the rewinding.
2. Only do the rewinding in cases that care about it, and have those cases opt in to it by some mechanism like hte ones I suggest in https://github.com/project-chip/connectedhomeip/pull/4048/files#r535734956 or something equivalent.  Of those options, I think I liked option 4 best, I think.

Reply:
Acked, currently, the msfBuff within sendMessage is relaying on upper layer to prevent it becoming dangling, this is not clear.

I updated with method 1 which uses an extra handle to retain the msgBuf for rewinding. Then we don't need to care from where the sendMessage get called.

Reply:
OK, but do all consumers want the rewinding behavior?

It's also a little weird to say that you will rewind a thing that you are taking ownership of.  This is relying on the fact that the consumer holds on to its own handle, and that the handles share state.  That's all true, but it seems like a somewhat weird API design.  Why this approach?

Reply:
I don't want to pass an extra parameter to SendMessage to indicate if the msgBuf need to be re-sent. For the packets do not need to be re-sent, they will be discarded anyway, nothing hurt to do the re-wind.

Adding the extra parameter will make the API more complicated and easier to make mistake.

Reply:
I want to keep all Transport Layer APIs unchanged for CRMP since CRMP is running in the context of exchange per spec, this design is also used in Weave, the lower level transport is only responsible for encoding/decoding, encrypting/decrypting. All re-transmission logics are transparent to lower level transport.

Reply:
> For the packets do not need to be re-sent, they will be discarded anyway

You don't know that.  A consumer could be doing whatever it wants to with the buffers.  Have you at least audited current consumers to ensure they are ok with the rewinding behavior?

> I want to keep all Transport Layer APIs unchanged for CRMP 

But we're not keeping them unchanged.  We're changing the behavior (from not rewinding to rewinding).  So the API change is there in terms of the method behavior; the only thing not changing is the method signature.   And in particular, this is changing the behavior for non-CRMP consumers too, necessitating that audit I asked about above.

Reply:
And in particular, I completely understand wanting retransmission logic to be in the higher-level CRMP code.  But since the encryption is in lower-level code, and since the lower-level code is what determines the initial packet buffer state for the send (after encryption and MAC), it needs to provide sufficient information to the higher-level code to do a retransmit.  Right now it does so in a very implicit way, by setting the state of a packet buffer via a handle and assuming that the higher level code has a handle to the same packet buffer, so the state is communicated via that shared packet buffer.  What I'm saying is that I would really prefer a much more explicit API that says "here is the state of the thing we sent", so higher levels can use that in whatever way they want.

Reply:
Yes, currently we don't have any case to re-use the buffer after SendMessage, CRMP is the first case which need to preserve the buffer for re-send. start and length of msgBuf are adjusted for several times during sending process even without rewinding.

I understand your rationality to notify the caller that if the msgBuf is re-winded during sending, I prefer to keep the API as simple as possible and harder to make mistake. Do you think we can just add an extra boolean para to current API, like:

CHIP_ERROR SendMessage(NodeId peerNodeId, System::PacketBufferHandle msgBuf, bool needsRewind);

Then only the packets sent from retransTable have needsRewind set to true?

Reply:
> start and length of msgBuf are adjusted for several times during sending process even without rewinding

Yes, I understand that, and that's fine behavior if the semantics of this method are "you have handed over ownership of this thing and won't touch it anymore".  But that's not the semantics after these changes; the new setup explicitly envisions some consumers continuing to use the same buffer even after what looks like an ownership handoff.

> Do you think we can just add an extra boolean para to current API, like:

We could.  This is option 3 from my list.  Though if the intent is to combine that with the "retain across the send" behavior, it's less error-prone than option 3 on its own, in that the code is not _relying_ on a ref in the caller that it has no way to enforce.

But why is this option preferable to option 4 from my list?  It ends up doing more buffer refcounting (if we assume that we will do an extra retain around the send).  What are the benefits over option 4?

Reply:
I originally prefer not changing the SecureSessionMgr API for CRMP, if we have to, then the change should be simple and minimal. 

Option 4 seems more complicated and I am not sure I get it correctly, do you means change this API something like:

CHIP_ERROR SecureSessionMgr::SendMessage(NodeId peerNodeId, System::PacketBufferHandle msgBuf, System::PacketBufferHandle * retainBuf)

If the retainBuf is not null, then use this one to retain the msgBuf and do the re-wind?

Reply:
I think we disagree on what constitutes a "minimal" change.  In my mind, a change that does not change the _signature_ but changes the behavior for the same signature, is still a change.

> If the retainBuf is not null, then use this one to retain the msgBuf and do the re-wind?

Yes, precisely.  And note that this would let us easily do the suggestion in https://github.com/project-chip/connectedhomeip/pull/4048#discussion_r537823647 which has this exact API shape, except for the type of the optional pointer argument.

Reply:
@bzbarsky-apple @rwalker-apple 

We have comprehensive discussion on various options, I think we should be able to make the final choice now.  Here is my proposal based on Boris's Option 4.

1. Add extra para System::PacketBufferHandle * retainBuf to the private SendMessage API: 
CHIP_ERROR SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId, System::PacketBufferHandle msgBuf, System::PacketBufferHandle * retainBuf, uint32_t & msgId, uint32_t & payloadLength, bool isResend);

If retainBuf is not null, this para will be used to retain the msgBuf and do the re-wind.

2. The public SendMessage API remain unchanged, for existing consumers of this API, no change on both API and internal implementation since retainBuf is set to null and isResend is set to false in the implementation of this API.

With this proposal, all recurred changes in TransportLayer will be limited to handling the messages sent from new added CRMP only.

Please lets me know if you are ok with this plan.

Reply:
Just to make sure I understand the tradeoffs behind that API shape, this requires `ReliableMessageManager` to be a friend of `SecureSessionMgr` to avoid exposing the no-encryption API too widely.  Is the reason that @rwalker-apple 's proposal is not being implemented as the public API the fact that we add to the retrans table before encryption when doing the initial send, so the retrans table API (`SendFromRetransTable`) involves a table entry and a "is this a resend" boolean?

I'd like to propose the following tweaks, if so:

1. The public API on `ReliableSessionManager` should not include `SendFromRetransTable` or `AddToRetransTable`.  Those are implementation details of reliable sessions.  The public API should just take the buffer to send, headers, a pointer to the `ReliableMessageContext` if needed.
2. Table entries store a `SecurePacketBufferHandle` (name to be determined), which indicates they are encrypted.
3. The public API on `ReliableSessionManager` adds a table entry but does not set its `SecurePacketBufferHandle` to anything yet.  It then calls a public `SendMessage` on the `SecureSessionMgr` and hands it the buffer to send and the pointer to the table entry's `SecurePacketBufferHandle`.  On failure, it cleans up the table entry as needed; this part of the code would be shared with `SendFromRetransTable`, e.g. via a helper function.
4. The public API on `SecureSessionMgr` has the API shape @rwalker-apple suggested.
5. There is a private API on `SecureSessionMgr` which looks like item 1 from your comment and is called by both public APIs.  The public API that takes `SecurePacketBufferHandle` does: `SendMessage(...., SecurePacketBufferHandle &already_encrypted) { SendMessage(...., std::move(already_encrypted), &already_encrypted, ....  , true)`.
6. The callsite of `SendFromRetransTable` in `ReliableMessageManager::ExecuteActions` remains, and `SendFromRetransTable` calls the API that takes `SecurePacketBufferHandle`.

This avoids the need for any friend declarations, ends up with a public API that can do resends of encrypted data only, enforced by the type system (which might turn out useful for group messaging too, not just for CRMP), and allows pretty decent code sharing between the send and resend code paths inside `SecureSessionMgr` as now and I suspect decent code sharing in `ReliableSessionManager` as well, if we use some helper functions for the shared bits of `SendFromRetransTable` and the initial "add to table and send" codepath.

Thoughts?

Reply:
"
Just to make sure I understand the tradeoffs behind that API shape, this requires ReliableMessageManager to be a friend of SecureSessionMgr to avoid exposing the no-encryption API too widely. Is the reason that @rwalker-apple 's proposal is not being implemented as the public API the fact that we add to the retrans table before encryption when doing the initial send, so the retrans table API (SendFromRetransTable) involves a table entry and a "is this a resend" boolean?
"
Yes, we need to avoid exposing any security hole to allow the external app to overpass the encryption process. retrans table does not contains "is this a resend" boolean, this is one of parameter in private SendMessage, and only get set to true in one case, the message is re-sent from retrans-table (not in the first send).

@bzbarsky-apple It seems your proposal will lead to revamp of current design, I am not sure I am fully get it. Could you list the new SendMessage APIs from your proposal?

Reply:
@bzbarsky-apple 

I compared the Option 3 and Option 4.

For Option 4, if we reuse  &entry->msgBuf as the retainBuf, we don't need to retain the buffer again within SendMessage since the buffer has already been retained as long as it is sent from retransTable. 

// Retain entry->msgBuf in mRetransTable before sending for following retransmition
err = mSessionMgr->SendMessage(entry->payloadHeader, entry->peerNodeId, entry->msgBuf.Retain(), msgId, payloadLength,  isResend);

So Option 3 use parameter 'boolean needsRewind' to tell SendMessage if to rewind the buffer, and Option 4 use parameter 'System::PacketBufferHandle *retainBuf' to tell SendMessage if to rewind the buffer. Is my understanding right?

Reply:
> Could you list the new SendMessage APIs from your proposal?

Something like this
```
class EncryptedPacketBufferHandle : public PacketBufferHandle {
public:
  explicit EncryptedPacketBufferHandle(EncryptedPacketBufferHandle&& aBuffer) :
    PacketBufferHandle(std::move(aBuffer))
  {}

  void operator=(EncryptedPacketBufferHandle&& aBuffer) {
    PacketBufferHandle::operator=(std::move(aBuffer));
  }

private:
  // Allow SecureSessionMgr to assign or construct us from a PacketBuffer
  friend class SecureSessionMgr;

  explicit EncryptedPacketBufferHandle(PacketBufferHandle&& aBuffer) :
    PacketBufferHandle(std::move(aBuffer))
  {}

  void operator=(PacketBufferHandle&& aBuffer) {
    PacketBufferHandle::operator=(std::move(aBuffer));
  }
};
...
class SecureSessionMgr {
public:
  CHIP_ERROR SendMessage(NodeId peerNodeId, System::PacketBufferHandle msgBuf);
  // Last arg could be some other struct that holds EncryptedPacketBufferHandle,
  // message id, payload length.
  CHIP_ERROR SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId, System::PacketBufferHandle msgBuf, RetransTableEntry * retransEntry = nullptr);
  // Ideally this next overload would _just_ take a buffer, since there's no
  // real need to set up all the headers all over again.  But we'd need to add
  // transport-level APIs for sending a buffer that has the header already in
  // it, so do it this way for now and file a followup to add the requisite
  // APIs.  bufferRetainSlot must be non-null.
  CHIP_ERROR SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId,  EncryptedPacketBufferHandle msgBuf, uint32_t msgId, uint32_t payloadLength, EncryptedPacketBufferHandle* bufferRetainSlot);

private:
   // Helper called by the other SendMessage overloads.  bufferRetainSlot may be
   // non-null to indicate that we want buffer rewinding and retaining into that
   // slot.
  CHIP_ERROR SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId, System::PacketBufferHandle msgBuf, EncryptedPacketBufferHandle* bufferRetainSlot, uint32_t & msgId,
			 uint32_t & payloadLength);
}
...

CHIP_ERROR SecureSessionMgr::SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId, System::PacketBufferHandle msgBuf, RetransTableEntry * retransEntry)
{
  if (!retransEntry) {
    uint32_t msgId = 0;
    uint32_t length = 0;
    return SendMessage(payloadHeader, peerNodeId, std::move(msgBuf), nullptr, msgId, length);
  }

  return SendMessage(payloadHeader, peerNodeId, std::move(msgBuf), *retransEntry->msgBuf, retransEntry->msgId, retransEntry->payloadLen);
}

CHIP_ERROR SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId,  EncryptedPacketBufferHandle msgBuf, uint32_t msgId, uint32_t payloadLength, EncryptedPacketBufferHandle* bufferRetainSlot)
{
  return SendMessage(payloadHeader, peerNodeId, std::move(msgBuf), bufferRetainSlot, msgId, payloadLen);  
}
```

As the comments say, it's not ideal; ideally we would just resend exactly the data that lives in the retained buffer, instead of rebuilding the headers, but it's a minimal-ish change from this PR, and gets us closer to doing that in the future, I think.

> we don't need to retain the buffer again within SendMessage since the buffer has already been retained as long as it is sent from retransTable

You move out of the retransTable entry on send, but the rewinding/retaining process puts it back into that entry, is the way I was thinking of it.  Using my API above, the callsite when doing a retransmit would look like this:
```
 err = mSessionMgr->SendMessage(entry->payloadHeader, entry->peerNodeId, std::move(entry->msgBuf), entry->msgId, entry->payloadLen, &entry->msgBuf);
```
and the initial CRMP send callsite would look like this, after adding an entry but not setting up its msgBuf yet:
```
 err = mSessionMgr->SendMessage(entry->payloadHeader, entry->peerNodeId, std::move(msgBuf), entry);
```

> So Option 3 use parameter 'boolean needsRewind' to tell SendMessage if to rewind the buffer, and Option 4 use parameter 'System::PacketBufferHandle *retainBuf' to tell SendMessage if to rewind the buffer.

The key part is that in option 4 the rewinding code is also doing the retain (via `*bufferRetainSlot = buf.Retain()` before it hands `buf` off to the transport layer), so it can guarantee that the retain happens.  Option 3 either has the safety issue of allowing callers to not have retained the buffer and then messing with its state, or does an extra retain for callers that don't actually care about retaining, afaict.

Reply:
I guess we could add the message id and payload length to `EncryptedPacketBufferHandle` and then the initial send API could just be:
```
  CHIP_ERROR SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId, System::PacketBufferHandle msgBuf, EncryptedPacketBufferHandle * bufferRetainSlot = nullptr);
```
and the resend API could be:
```
  CHIP_ERROR SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId,  EncryptedPacketBufferHandle msgBuf, EncryptedPacketBufferHandle* bufferRetainSlot);
```
or so, to avoid inventing another struct or passing around retrans table entries.

Reply:
@bzbarsky-apple Thanks for your comments, I am piloting your proposal now to see how much of the change, I try to find a way to re-use the current logic in ReliableMessageManager since its core logic is depending on AddToRetransTable and SendFromRetransTable. If we remove them, we still need to figure out if there is slot in retransTable to send, when to calculate nextRetransTimeTick, when to trigger the re-transmit timer and when to increase the reference counting of exchange

Reply:
@bzbarsky-apple 

I hit the following two issues as I proceed with the refactor.

1.  Now we introduce an new API to resend encrypted package. 
CHIP_ERROR SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId,  EncryptedPacketBufferHandle msgBuf, EncryptedPacketBufferHandle* bufferRetainSlot);

The user can use this API to overpass the encryption process just by moving a PacketBufferHandle to EncryptedPacketBufferHandle and pass to this function, right?

2. Now the first message is not sent from retrainsTable, we need to first send out this message, and then add the returned bufferRetainSlot to retrainsTable, and triggering the resend timer,. What if  we fail to add the returned bufferRetainSlot to retrainsTable after we send out the first message?  I think one solution is to add another pre-check function to make sure there is a free slot in the retransTable before we send out the first message.

Reply:
> The user can use this API to overpass the encryption process just by moving a PacketBufferHandle to EncryptedPacketBufferHandle and pass to this function, right?

No, because the relevant APIs on `EncryptedPacketBufferHandle` are private, and the only friend is `SecureSessionMgr`.  The only public APIs on `EncryptedPacketBufferHandle` are the move constructor and move assignment operator (plus the implicitly public destructor).

> we need to first send out this message, and then add the returned bufferRetainSlot to retrainsTable, and trigger the resend timer

The way I was envisioning this working was:

1. Add an entry to the retrans table; its `EncryptedPacketBufferHandle` holds null for the moment.
2. If the addition fails, fail out.
3. Do the first message send, passing it a pointer to the table entry's `EncryptedPacketBufferHandle`.  If the send fails, clean up the retrans table entry as needed.

I thought I had mentioned that above....

Reply:
Acked. Updated the PR with your suggested proposal.